File: dtx/runtime/kotlin/src/main/kotlin/org/llvm/dtx/KeyedArchiver.kt
package org.llvm.dtx

import java.io.ByteArrayOutputStream
import java.nio.charset.StandardCharsets

object KeyedArchiver {
    private val magic = byteArrayOf('D'.code.toByte(), 'T'.code.toByte(), 'X'.code.toByte(), 'K'.code.toByte(), 'A'.code.toByte(), 'R'.code.toByte(), '1'.code.toByte(), 0)

    private object Tag {
        const val NULL = 0
        const val BOOL = 1
        const val INT = 2
        const val UINT = 3
        const val DOUBLE = 4
        const val STRING = 5
        const val DATA = 6
        const val ARRAY = 7
        const val DICT = 8
    }

    fun archiveRoot(root: NSObject): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(magic)
        writeObject(out, root)
        return out.toByteArray()
    }

    fun unarchiveRoot(bytes: ByteArray): NSObject {
        if (bytes.size < magic.size || !bytes.copyOfRange(0, magic.size).contentEquals(magic))
            throw DtxException("Bad object archive magic.")
        val reader = Reader(bytes, magic.size)
        val root = reader.readObject()
        if (!reader.consumed)
            throw DtxException("Trailing bytes after object archive root.")
        return root
    }

    fun unarchiveRoot(bytes: ByteArray, offset: Int, length: Int): NSObject =
        unarchiveRoot(bytes.copyOfRange(offset, offset + length))

    private fun writeObject(out: ByteArrayOutputStream, value: NSObject) {
        when (value) {
            NSObject.Null -> out.write(Tag.NULL)
            is NSObject.BoolValue -> {
                out.write(Tag.BOOL)
                out.write(if (value.value) 1 else 0)
            }
            is NSObject.IntValue -> {
                out.write(Tag.INT)
                writeU64(out, value.value)
            }
            is NSObject.UIntValue -> {
                out.write(Tag.UINT)
                writeU64(out, value.value)
            }
            is NSObject.DoubleValue -> {
                out.write(Tag.DOUBLE)
                writeU64(out, java.lang.Double.doubleToRawLongBits(value.value))
            }
            is NSObject.StringValue -> {
                out.write(Tag.STRING)
                writeBytes(out, value.value.toByteArray(StandardCharsets.UTF_8))
            }
            is NSObject.DataValue -> {
                out.write(Tag.DATA)
                writeBytes(out, value.value)
            }
            is NSObject.ArrayValue -> {
                out.write(Tag.ARRAY)
                writeU32(out, value.value.size)
                value.value.forEach { writeObject(out, it) }
            }
            is NSObject.DictValue -> {
                out.write(Tag.DICT)
                val sorted = value.value.toSortedMap()
                writeU32(out, sorted.size)
                sorted.forEach { (key, item) ->
                    writeBytes(out, key.toByteArray(StandardCharsets.UTF_8))
                    writeObject(out, item)
                }
            }
        }
    }

    private fun writeBytes(out: ByteArrayOutputStream, bytes: ByteArray) {
        writeU32(out, bytes.size)
        out.write(bytes)
    }

    private class Reader(private val bytes: ByteArray, private var offset: Int = 0) {
        val consumed: Boolean
            get() = offset == bytes.size

        fun readObject(): NSObject =
            when (val tag = readByte().toInt() and 0xff) {
                Tag.NULL -> NSObject.Null
                Tag.BOOL -> NSObject.bool(readByte().toInt() != 0)
                Tag.INT -> NSObject.int(readU64())
                Tag.UINT -> NSObject.uint(readU64())
                Tag.DOUBLE -> NSObject.double(java.lang.Double.longBitsToDouble(readU64()))
                Tag.STRING -> NSObject.string(String(readBytes(), StandardCharsets.UTF_8))
                Tag.DATA -> NSObject.data(readBytes())
                Tag.ARRAY -> readArray()
                Tag.DICT -> readDict()
                else -> throw DtxException("Unknown object archive tag: $tag")
            }

        private fun readArray(): NSObject {
            val count = readU32()
            val values = ArrayList<NSObject>(count)
            repeat(count) { values.add(readObject()) }
            return NSObject.array(values)
        }

        private fun readDict(): NSObject {
            val count = readU32()
            val values = LinkedHashMap<String, NSObject>()
            repeat(count) {
                val key = String(readBytes(), StandardCharsets.UTF_8)
                values[key] = readObject()
            }
            return NSObject.dict(values)
        }

        private fun readByte(): Byte {
            if (offset >= bytes.size)
                throw DtxException("Unexpected EOF while reading object archive.")
            return bytes[offset++]
        }

        private fun readU32(): Int {
            if (offset + 4 > bytes.size)
                throw DtxException("Unexpected EOF while reading u32.")
            val value = u32At(bytes, offset)
            offset += 4
            return value
        }

        private fun readU64(): Long {
            if (offset + 8 > bytes.size)
                throw DtxException("Unexpected EOF while reading u64.")
            val value = u64At(bytes, offset)
            offset += 8
            return value
        }

        private fun readBytes(): ByteArray {
            val length = readU32()
            if (offset + length > bytes.size)
                throw DtxException("Object archive byte payload extends past buffer.")
            val out = bytes.copyOfRange(offset, offset + length)
            offset += length
            return out
        }
    }
}

internal fun writeU32(out: ByteArrayOutputStream, value: Int) {
    out.write(value and 0xff)
    out.write((value ushr 8) and 0xff)
    out.write((value ushr 16) and 0xff)
    out.write((value ushr 24) and 0xff)
}

internal fun writeU64(out: ByteArrayOutputStream, value: Long) {
    var current = value
    repeat(8) {
        out.write((current and 0xffL).toInt())
        current = current ushr 8
    }
}

internal fun u32At(bytes: ByteArray, offset: Int): Int =
    (bytes[offset].toInt() and 0xff) or
        ((bytes[offset + 1].toInt() and 0xff) shl 8) or
        ((bytes[offset + 2].toInt() and 0xff) shl 16) or
        ((bytes[offset + 3].toInt() and 0xff) shl 24)

internal fun u64At(bytes: ByteArray, offset: Int): Long {
    var value = 0L
    for (i in 7 downTo 0)
        value = (value shl 8) or (bytes[offset + i].toLong() and 0xffL)
    return value
}

File: dtx/runtime/csharp/src/Dtx/KeyedArchiver.cs
using System.Text;

namespace LLVM.DTX;

public static class KeyedArchiver
{
    private static readonly byte[] Magic = { (byte)'D', (byte)'T', (byte)'X', (byte)'K', (byte)'A', (byte)'R', (byte)'1', 0 };

    private enum Tag : byte
    {
        Null = 0,
        Bool = 1,
        Int = 2,
        UInt = 3,
        Double = 4,
        String = 5,
        Data = 6,
        Array = 7,
        Dict = 8,
    }

    public static byte[] ArchiveRoot(NSObject root)
    {
        using var stream = new MemoryStream();
        stream.Write(Magic, 0, Magic.Length);
        WriteObject(stream, root);
        return stream.ToArray();
    }

    public static NSObject UnarchiveRoot(byte[] bytes) =>
        UnarchiveRoot(bytes, 0, bytes.Length);

    public static NSObject UnarchiveRoot(byte[] bytes, int offset, int length)
    {
        if (length < Magic.Length)
            throw new DtxException("Bad object archive magic.");
        var archive = BinaryUtil.Slice(bytes, offset, length);
        if (!BinaryUtil.StartsWith(archive, Magic))
            throw new DtxException("Bad object archive magic.");
        var reader = new Reader(archive, Magic.Length);
        var root = reader.ReadObject();
        if (!reader.Consumed)
            throw new DtxException("Trailing bytes after object archive root.");
        return root;
    }

    private static void WriteObject(Stream stream, NSObject value)
    {
        stream.WriteByte((byte)(value.Kind switch
        {
            NSObjectKind.Null => Tag.Null,
            NSObjectKind.Bool => Tag.Bool,
            NSObjectKind.Int => Tag.Int,
            NSObjectKind.UInt => Tag.UInt,
            NSObjectKind.Double => Tag.Double,
            NSObjectKind.String => Tag.String,
            NSObjectKind.Data => Tag.Data,
            NSObjectKind.Array => Tag.Array,
            NSObjectKind.Dict => Tag.Dict,
            _ => throw new DtxException("Unsupported NSObject kind."),
        }));

        switch (value.Kind)
        {
            case NSObjectKind.Null:
                break;
            case NSObjectKind.Bool:
                stream.WriteByte((bool)value.Value! ? (byte)1 : (byte)0);
                break;
            case NSObjectKind.Int:
                WriteU64(stream, unchecked((ulong)(long)value.Value!));
                break;
            case NSObjectKind.UInt:
                WriteU64(stream, (ulong)value.Value!);
                break;
            case NSObjectKind.Double:
                WriteU64(stream, BinaryUtil.DoubleToUInt64Bits((double)value.Value!));
                break;
            case NSObjectKind.String:
                WriteBytes(stream, Encoding.UTF8.GetBytes((string)value.Value!));
                break;
            case NSObjectKind.Data:
                WriteBytes(stream, (byte[])value.Value!);
                break;
            case NSObjectKind.Array:
                var array = (NSObject[])value.Value!;
                WriteU32(stream, checked((uint)array.Length));
                foreach (var item in array)
                    WriteObject(stream, item);
                break;
            case NSObjectKind.Dict:
                var dict = (IReadOnlyDictionary<string, NSObject>)value.Value!;
                WriteU32(stream, checked((uint)dict.Count));
                foreach (var item in dict.OrderBy(kv => kv.Key, StringComparer.Ordinal))
                {
                    WriteBytes(stream, Encoding.UTF8.GetBytes(item.Key));
                    WriteObject(stream, item.Value);
                }
                break;
        }
    }

    private static void WriteBytes(Stream stream, byte[] bytes)
    {
        WriteU32(stream, checked((uint)bytes.Length));
        stream.Write(bytes, 0, bytes.Length);
    }

    private static void WriteU32(Stream stream, uint value)
    {
        BinaryUtil.WriteUInt32LE(stream, value);
    }

    private static void WriteU64(Stream stream, ulong value)
    {
        BinaryUtil.WriteUInt64LE(stream, value);
    }

    private sealed class Reader
    {
        private readonly byte[] _bytes;
        private int _offset;

        public Reader(byte[] bytes, int offset)
        {
            _bytes = bytes;
            _offset = offset;
        }

        public bool Consumed => _offset == _bytes.Length;

        public NSObject ReadObject()
        {
            var tag = (Tag)ReadByte();
            return tag switch
            {
                Tag.Null => NSObject.Null,
                Tag.Bool => NSObject.Bool(ReadByte() != 0),
                Tag.Int => NSObject.Int(unchecked((long)ReadU64())),
                Tag.UInt => NSObject.UInt(ReadU64()),
                Tag.Double => NSObject.Double(BinaryUtil.UInt64BitsToDouble(ReadU64())),
                Tag.String => NSObject.String(Encoding.UTF8.GetString(ReadBytes())),
                Tag.Data => NSObject.Data(ReadBytes()),
                Tag.Array => ReadArray(),
                Tag.Dict => ReadDict(),
                _ => throw new DtxException("Unknown object archive tag."),
            };
        }

        private NSObject ReadArray()
        {
            var count = ReadU32();
            var values = new NSObject[count];
            for (var i = 0; i < values.Length; i++)
                values[i] = ReadObject();
            return NSObject.Array(values);
        }

        private NSObject ReadDict()
        {
            var count = ReadU32();
            var values = new Dictionary<string, NSObject>(StringComparer.Ordinal);
            for (uint i = 0; i < count; i++)
            {
                var key = Encoding.UTF8.GetString(ReadBytes());
                values[key] = ReadObject();
            }
            return NSObject.Dict(values);
        }

        private byte ReadByte()
        {
            if (_offset >= _bytes.Length)
                throw new DtxException("Unexpected EOF while reading object archive.");
            return _bytes[_offset++];
        }

        private uint ReadU32()
        {
            if (_offset + 4 > _bytes.Length)
                throw new DtxException("Unexpected EOF while reading u32.");
            var value = BinaryUtil.ReadUInt32LE(_bytes, _offset);
            _offset += 4;
            return value;
        }

        private ulong ReadU64()
        {
            if (_offset + 8 > _bytes.Length)
                throw new DtxException("Unexpected EOF while reading u64.");
            var value = BinaryUtil.ReadUInt64LE(_bytes, _offset);
            _offset += 8;
            return value;
        }

        private byte[] ReadBytes()
        {
            var length = checked((int)ReadU32());
            if (_offset + length > _bytes.Length)
                throw new DtxException("Object archive byte payload extends past buffer.");
            var bytes = BinaryUtil.Slice(_bytes, _offset, length);
            _offset += length;
            return bytes;
        }
    }
}
